GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -55,29 +56,32 @@ public void Dispose()
         private void TimerCallback(object unusedState)
         {
             string errorMessage = null;
-
             InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
-            ProductUpgrader productUpgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
-            if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
+            IProductUpgrader productUpgrader = ProductUpgrader.CreateUpgrader(this.tracer, out errorMessage);
+
+            if (productUpgrader != null)

```

Comment:
Is it an error if the factory cannot instantiate an upgrader?

Reply:
Oh, I see. We're tracing later if anything has set `errorMessage`?

Reply:
This confused me as well.  Can this be updated to use the ` bool TryXXX` pattern (i.e. `TryCreateUpgrader`)

Reply:
>>Is it an error if the factory cannot instantiate an upgrader?

Yes.

>>We're tracing later if anything has set errorMessage?

Right.